[FG:InPlacePodVerticalScaling] Use latest status from runtime at resizing

hshiina opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind bug

#### What this PR does / why we need it:
In order to solve a couple of problems caused by requesting pod resizing consecutively, this PR uses the latest pod status at the following work in resizing:
- Update pod status
  If resource requests are requested to be updated, `allocatedResources` and `resize` in the pod status is updated. This update should base on the latest status. Otherwise, the pod status will be overwritten with an outdated status in `pod`.
- Compute pod resize actions
  `computePodResizeAction()` refers to the status in `pod` which is replaced in `handlePodResourcesResize()`. In order to determine actions based on the latest status, the pod status should be updated with the latest status.

`apiPodStatus`, which is copied to `pod.Status`, is generated from the latest pod status that a pod worker gets from the PLEG cache.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #125205

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed a bug that it takes one more minute when a pod with the RestartContainer resize policy is requested to be resized a couple of times quickly. Also fixed a bug that a pod status is sometimes regressed to outdated information in the same case.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

esotsal commented:
/ok-to-test

esotsal commented:
/test pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2

esotsal commented:
/test pull-kubernetes-node-e2e-alpha-ec2

hshiina committed:
Use latest status from runtime at resizing

hshiina commented:
/test pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2
/test pull-kubernetes-node-e2e-alpha-ec2

esotsal commented:
/triage accepted
/priority important-longterm

hshiina commented:
/test pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2

esotsal commented:
/test pull-kubernetes-node-e2e-alpha-ec2

hshiina committed:
e2e: Add consecutive resize test

esotsal commented:
/test pull-kubernetes-node-kubelet-serial-podresize

kannon92 commented:
/cc @tallclair

esotsal commented:
Thanks @hshiina for this patch, i think it solved the problem checking [the test run](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-node-kubelet-serial-podresize/1816134427252625408). Good job

tallclair commented:
/assign

AnishShah commented on test/e2e_node/pod_resize_test.go:
```diff
@@ -1295,6 +1302,570 @@ func doPodResizeErrorTests() {
 	}
 }
 
+func doConsecutivePodResizeTests() {
```
I'm refactoring IPPR cluster and node e2e tests in https://github.com/kubernetes/kubernetes/pull/128143.

Can we wait until that PR is merged?

AnishShah reviewed (commented):
This PR also adds e2e tests asked in #109905? Can we add "Fixed #issue" in the first comment?

hshiina commented:
/test pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2
/test pull-kubernetes-node-kubelet-serial-podresize

AnishShah commented:
/test pull-kubernetes-node-e2e-containerd

AnishShah commented on test/e2e/common/node/pod_resize.go:
```diff
@@ -1012,6 +1013,565 @@ func doPodResizeErrorTests(f *framework.Framework) {
 
+func doConsecutivePodResizeTests(f *framework.Framework) {
+	var podClient *e2epod.PodClient
+	ginkgo.BeforeEach(func() {
+		podClient = e2epod.NewPodClient(f)
+	})
```
nit: Is it possible to initialize `podClient` inside each testcase? I refactored this for other tests in my PR #128143.

AnishShah commented on test/e2e/common/node/pod_resize.go:
```diff
@@ -1012,6 +1013,565 @@ func doPodResizeErrorTests(f *framework.Framework) {
+			tStamp := strconv.Itoa(time.Now().Nanosecond())
+			e2epod.InitDefaultResizePolicy(tc.containers)
+			for _, e := range tc.expected {
+				if len(e) > 0 {
+					e2epod.InitDefaultResizePolicy(e)
+				}
```
I don't think we need to check the length. `InitDefaultResizePolicy` will handle slices with length 0.

```suggestion
				e2epod.InitDefaultResizePolicy(e)
```

AnishShah commented on test/e2e/common/node/pod_resize.go:
```diff
@@ -1012,6 +1013,565 @@ func doPodResizeErrorTests(f *framework.Framework) {
+		isSynchronous    bool
+		name             string
+		containers       []e2epod.ResizableContainerInfo
+		containerPatches [][]v1.Container
```
Similar to other tests, it might be better to use patch string. We don't want to wait for JSON marshaling between consecutive resize request (especially for asynchronous testcase)

AnishShah reviewed (commented):
Thank you for refactoring the tests.

hshiina commented on test/e2e/common/node/pod_resize.go:
```diff
@@ -1012,6 +1013,565 @@ func doPodResizeErrorTests(f *framework.Framework) {
+				e2epod.VerifyPodResources(resizedPod, tc.expected[i])
+
+				ginkgo.By("verifying pod allocations after resize")
+				err = e2epod.VerifyPodAllocations(resizedPod, tc.expected[i])
```
I noticed `VerifyPodAllcation()` was already done in `WaitForPodResizeActuation()`. I will remove this verification.

hshiina committed:
e2e: Fix consecutive resize test

Especially, add a signal handler to the command that containers
execute in order to restart containers quickly.

hshiina commented:
/test pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2
/test pull-kubernetes-node-kubelet-serial-podresize

tallclair commented:
I think this is superseded by https://github.com/kubernetes/kubernetes/pull/128269

/close
